src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,781 @@
+
+      C(i,i) = rgD[i] = rgInitialStds[i]*std::sqrt(N/trace);
+      C(i,i) *= C(i,i);
+      pc[i] = ps[i] = double(0);
+    }
+    minEW = rgD.min();
+    minEW = minEW*minEW;
+    maxEW = rgD.max();
+    maxEW = maxEW*maxEW;
+
+    maxdiagC = arma::max(C.diag());
+    mindiagC = arma::min(C.diag());
+
+    for (int i = 0; i < N; ++i)
+      xmean[i] = xold[i] = xstart[i];
+    
+    if (typicalXcase)
+      for (int i = 0; i < N; ++i)
+        xmean[i] += sigma*rgD[i]*rand.gauss();
+
+    for(int i=0; i<lambda; i++) func[i] = publicFitness[i];
+  }
+
+  /**
+   * The search space vectors have a special form: they are arrays with N+1
+   * entries. Entry number -1 is the dimension of the search space N.
+   * @return A pointer to a "population" of lambda N-dimensional multivariate
+   * normally distributed samples.
+   */
+   template<typename funcType>
+   void CMAES<funcType>::samplePopulation()
+  {
+    bool diag = diagonalCov == 1 || diagonalCov >= gen;
+
+    // calculate eigensystem
+    if (!eigensysIsUptodate)
+    {
+      if (!diag)
+        updateEigensystem(false);
+      else
+      {
+        for (int i = 0; i < N; ++i)
+          rgD[i] = std::sqrt(C(i,i));
+        minEW = rgD.min();
+        minEW *= minEW;
+        maxEW = rgD.max();
+        maxEW *= maxEW;
+        eigensysIsUptodate = true;
+      }
+    }
+
+    for (int iNk = 0; iNk < lambda; ++iNk)
+    { // generate scaled random vector D*z
+      for (int i = 0; i < N; ++i)
+        if (diag)
+          population(iNk,i) = xmean[i] + sigma*rgD[i]*rand.gauss();
+        else
+          tempRandom[i] = rgD[i]*rand.gauss();
+      if (!diag)
+        for (int i = 0; i < N; ++i) // add mutation sigma*B*(D*z)
+        {
+          double sum = 0.0;
+          for (int j = 0; j < N; ++j)
+            sum += B(i,j)*tempRandom[j];
+          population(iNk , i) = xmean[i] + sigma*sum;
+        }
+    }
+
+    if (state == UPDATED || gen == 0)
+      ++gen;
+    state = SAMPLED;
+
+  }
+
+  /**
+   * Core procedure of the CMA-ES algorithm. Sets a new mean value and estimates
+   * the new covariance matrix and a new step size for the normal search
+   * distribution.
+   * @param fitnessValues An array of \f$\lambda\f$ function values.
+   * @return Mean value of the new distribution.
+   */
+   template<typename funcType>
+  void CMAES<funcType>::updateDistribution(arma::vec fitnessValues)
+  {
+
+    bool diag = diagonalCov == 1 || diagonalCov >= gen;
+
+    assert(state != UPDATED && "updateDistribution(): You need to call "
+          "samplePopulation() before update can take place.");
+
+    if (state == SAMPLED) // function values are delivered here
+      countevals += lambda;
+    else std::cout<<  "updateDistribution(): unexpected state" << std::endl;
+
+    // assign function values
+    for (int i = 0; i < lambda; ++i)
+      population(i,N) = functionValues[i] = fitnessValues[i];
+
+    // Generate index
+    sortIndex(fitnessValues, index, lambda);
```

Comment:
@zoq index must had to be of the type uvec for this. changed as well.. thanks
done